tests/test_qnode.py

```diff
@@ -1398,6 +1348,488 @@ def circuit(a):
             circuit.jacobian([1.0], method="A")
 
 
+class TestMetricTensor:
+    """Tests for metric tensor subcircuit construction and evaluation"""
+
+    def test_no_generator(self):
+        """Test exception is raised if subcircuit contains an
+        operation with no generator"""
+        dev = qml.device('default.qubit', wires=1)
+
+        def circuit(a):
+            qml.Rot(a, 0, 0, wires=0)
+            return qml.expval(qml.PauliX(0))
+
+        circuit = qml.QNode(circuit, dev)
+
+        with pytest.raises(QuantumFunctionError, match="has no defined generator"):
+            circuit.construct_metric_tensor([1])
+
+    def test_generator_no_expval(self, monkeypatch):
+        """Test exception is raised if subcircuit contains an
+        operation with generator that corresponds to no expectation value"""
+        dev = qml.device('default.qubit', wires=1)
+
+        def circuit(a):
+            qml.RX(a, wires=0)
+            return qml.expval(qml.PauliX(0))
+
+        circuit = qml.QNode(circuit, dev)
+
+        with monkeypatch.context() as m:
+            m.setattr('pennylane.RX.generator', [qml.RX, 1])
+
+            with pytest.raises(QuantumFunctionError, match="no corresponding expectation value"):
+                circuit.construct_metric_tensor([1])
+
+    def test_construct_subcircuit(self):
+        """Test correct subcircuits constructed"""
+        dev = qml.device('default.qubit', wires=2)
+
+        def circuit(a, b, c):
+            qml.RX(a, wires=0)
+            qml.RY(b, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.PhaseShift(c, wires=1)
+            return qml.expval(qml.PauliX(0))
+
+        circuit = qml.QNode(circuit, dev)
+
+        circuit.construct_metric_tensor([1, 1, 1])
+        res = circuit._metric_tensor_subcircuits
+
+        # first parameter subcircuit
+        assert len(res[(0,)]['queue']) == 0
+        assert isinstance(res[(0,)]['observable'][0], qml.PauliX)
+
+        # second parameter subcircuit
+        assert len(res[(1,)]['queue']) == 1
+        assert isinstance(res[(1,)]['queue'][0], qml.RX)
+        assert isinstance(res[(1,)]['observable'][0], qml.PauliY)
+
+        # third parameter subcircuit
+        assert len(res[(2,)]['queue']) == 3
+        assert isinstance(res[(2,)]['queue'][0], qml.RX)
+        assert isinstance(res[(2,)]['queue'][1], qml.RY)
+        assert isinstance(res[(2,)]['queue'][2], qml.CNOT)
+        assert isinstance(res[(2,)]['observable'][0], qml.Hermitian)
+        assert np.all(res[(2,)]['observable'][0].params[0] == qml.PhaseShift.generator[0])
+
+    def test_construct_subcircuit_layers(self):
+        """Test correct subcircuits constructed
+        when a layer structure exists"""
+        dev = qml.device('default.qubit', wires=3)
+
+        def circuit(params):
+            # section 1
+            qml.RX(params[0], wires=0)
+            # section 2
```

Comment:
One thing that is still unclear to me is: what exactly is a block/layer/section?

A related point is: from my earlier reading of the implementation, a block was determined by examining only the ordering and dependence amongst ops. Do we use the information that certain gates on the same subcircuit might commute, and should go in the same block?

Reply:
Ah! Good question.

When I originally coded this up, I had a block of code in `construct_metric_tensor` that first checked if _every_ pairwise combinations of expanded arrays commuted. If they did, it followed the block_diag branch. if not, it assumed the diagonal approximation. (The layer identification worked as it did now - it simply identified parametrized layers without worrying about observables).

This was a _massive_ bottleneck, and caused the construction to take a extraordinarily long amount of time. So I removed it, with the assumption that it is impossible to construct parametrized layers in PennyLane with generators that do not commute (since the only 4 parametrized gates with generators are Rx, Ry, Rz, and PhaseShift, and their generators all commute).

Reply:
So, to confirm: block diagonal structure is purely determined by the order of gates within layers, and not due to their actual values?

Reply:
> So, to confirm: block diagonal structure is purely determined by the order of gates within layers, and not due to their actual values?

Yep. Layer structure is calculated separately, and then `construct_metric_tensor` assumes that all observables in each layer commute.

The order of each gate _within_ the metric tensor block diagonal is determined by the parameter index, to match the manuscript.